### PR TITLE
6374: List of dirty leaves doesn't have to be sorted on VirtualNodeCache flush

### PR DESCRIPTION
**Pull request by artemananiev**
Fix summary:

* `VirtualNodeCache.dirtyLeaves()` is changed to sort or not sort the stream of dirty leaves. Unsorted streams are used for flushes, sorted streams are needed for hashing
* `VirtualNodeCache.dirtyHashes()` now always returns an unsorted stream of dirty hashes
* Even unsorted streams of hashes / leaves must not contain duplicate items. Previously, streams were always sorted, and then duplicates were removed using `dedupeByPath()` method. Now duplicates are removed differently. First, for all mutations in `dirtyLeaves` / `dirtyHashes`, "next" mutation is marked. These next mutations correspond to earlier cache versions, they are overridden by newer mutations and can be ignored. On the second pass, all marked mutations are filtered out
* `ConcurrentArray.sortedStream()` is renamed to `stream()`
* Since `VirtualDataSourceJasperDB` still expects the streams to be sorted, sorting is done right in its `saveRecords()` method

Fixes: https://github.com/hashgraph/hedera-services/issues/6374
Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>

### CODE REVIEW

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/jasperdb/VirtualDataSourceJasperDB.java [hunk 1]
```diff
@@ -1251,7 +1252,9 @@ private void writeLeaves(
 
             // iterate over leaf records
             final AtomicLong lastPath = new AtomicLong(INVALID_PATH);
-            leafRecordsToAddOrUpdate.forEach(leafRecord -> {
+            final Stream<VirtualLeafRecord<K, V>> sorderDirtyLeaves =
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/jasperdb/VirtualDataSourceJasperDB.java:1255]: typo: `sortedDirtyLeaves`?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java [hunk 2]
```diff
@@ -284,31 +314,31 @@ void add(T element) {
      * 		If this instance is not immutable.
      */
     @SuppressWarnings("unchecked")
-    Stream<T> sortedStream(Comparator<T> comparator) {
+    Stream<T> stream(final boolean sort, @Nullable final Comparator<T> comparator) {
         if (!immutable.get()) {
-            throw new IllegalStateException("You can not call sortedStream on a mutable ConcurrentArray");
+            throw new IllegalStateException("You can not call toArray on a mutable ConcurrentArray");
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java:319]: Why `toArray`? `stream`?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java [hunk 3]
```diff
@@ -284,31 +314,31 @@ void add(T element) {
...
         }
 
-        // A quick exit: if there are no elements, then return an empty stream.
+        // A quick exit: if there are no elements, then return an empty array
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java:322]: s/array/stream/ ?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java [hunk 4]
```diff
@@ -743,26 +733,36 @@ public VirtualLeafRecord<K, V> lookupLeafByPath(final long path, final boolean f
...
                 })
-                .filter(mutation -> dedupeLeafByPath(mutation, lastSeen))
-                .filter(mutation -> !mutation.deleted)
+                .filter(mutation -> {
+                    if (mutation.next != null) {
+                        mutation.next.setFiltered();
+                    }
+                    return true;
+                })
+                .filter(mutation -> !mutation.isFiltered())
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:758]: You can't use `setFiltered()` and `isFiltered()` is one pass: a record may be read from the stream and later be marked as filtered.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java [hunk 5]
```diff
@@ -928,21 +928,22 @@ public Hash lookupHashByPath(final long path, final boolean forModify) {
...
                 .filter(mutation -> mutation.key <= lastLeafPath)
-                .filter(mutation -> dedupeHashByPath(mutation, lastSeen))
-                .filter(mutation -> !mutation.deleted)
+                .filter(mutation -> {
+                    if (mutation.next != null) {
+                        mutation.next.setFiltered();
+                    }
+                    return true;
+                })
+                .filter(mutation -> !mutation.isFiltered())
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:946]: Ditto.

**Review by OlegMazurov — CHANGES_REQUESTED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java [hunk 6]
```diff
@@ -272,9 +273,38 @@ void add(T element) {
...
+     * Get a stream of all elements in this concurrent array. The elements will be sorted using the
+     * provided comparator, if not null, or in natural order.
+     *
+     * <p>This method can only be called on immutable instances. It may be called concurrently.
+     *
+     * @return A non-null sorted stream over all elements
+     * @throws IllegalStateException
+     * 		If this instance is not immutable
+     */
+    Stream<T> stream(@Nullable final Comparator<T> comparator) {
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java:298]: Is this method used outside tests?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java (hunk 4 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:758]: Should it be possible with reverse stream over the array?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java (hunk 2 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java:319]: Fixed

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java (hunk 3 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java:322]: Fixed

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java (hunk 4 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java:755]: OK, it's now done in two passes. First, all obsolete mutations are explicitly marked to filter, using `ConcurrentArray.parallelTraverse()` mechanism. Second, the resulting stream is filtered based on the marks.

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/jasperdb/VirtualDataSourceJasperDB.java (hunk 1 above)
Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/jasperdb/VirtualDataSourceJasperDB.java]: Fixed.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java (hunk 6 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java]: This method is removed in the latest fix version.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java [hunk 7]
```diff
@@ -272,43 +273,26 @@ void add(T element) {
...
-            throw new IllegalStateException("You can not call sortedStream on a mutable ConcurrentArray");
+            throw new IllegalStateException("You can not call stream() on a mutable ConcurrentArray");
         }
 
-        // A quick exit: if there are no elements, then return an empty stream.
+        // A quick exit: if there are no elements, then return an empty stream
         final int numberOfElements = this.elementCount.get();
         if (numberOfElements == 0) {
-            return Stream.empty();
+            return Stream.of();
```
Comment (imalygin) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java:292]: Just for my understanding. Is there any particular difference or it's as shorter form of the same thing?

Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java:292]: This change is not intentional. Let me revert it back to `empty()`

**Review by imalygin — DISMISSED** [review]

**Review by OlegMazurov — DISMISSED** [review]

**Review by imalygin — APPROVED** [review]

**Review by OlegMazurov — DISMISSED** [review]

**Review by OlegMazurov — APPROVED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java [hunk 8]
```diff
@@ -392,4 +376,40 @@ boolean add(T element) {
...
+     * Utility class to provide Stream support on top of concurrent arrays.
+     *
+     * @param <T>
+     */
+    private static class ConcurrentArraySpliterator<T> extends AbstractSpliterator<T> {
+        private SubArray<T> arr;
+        private int arrIndex = 0;
+
+        ConcurrentArraySpliterator(final int size, final SubArray<T> head) {
+            super(size, Spliterator.SIZED | Spliterator.NONNULL | Spliterator.IMMUTABLE);
```
Comment (jsync-swirlds) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/ConcurrentArray.java:390]: Should include `Spliterator.CONCURRENT` here so this ConcurrentArray can be used with `parallelStream`

**Pull request merged**